sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memory_alloc.asciidoc

```diff
@@ -0,0 +1,755 @@
+with.
+
+_Throws_: An exception with the `errc::invalid` error code if `ptr` is not the
+address of a memory allocation allocated to a memory pool.
+
+[_Note:_ Accessing the memory at the address of `ptr` after the asynchronous
+free command has completed execution is undefined behavior. _{endnote}_]
+
+
+=== Memory pool lifetimes
+
+The lifetime of memory allocated via a memory pool is as follows.
+
+When an asynchronous malloc command is executed it will first look to the
+memory pool to opportunistically allocate memory already available there. The
+SYCL RT may also, when possible, look at all asynchronous malloc and free
+commands and their dependencies in order to more efficiently schedule memory
+allocations. If the memory pool does not have the requirement memory available
+to allocate for the asynchronous malloc command it will then look to allocate
+from the source to the memory pool to provide the required memory.
+
+When an asynchronous free command is executed it will free up memory in the 
+memory pool to be used for a later asynchonrous malloc command. Memory is not
+freed back to the source at this point.
+
+When a SYCL queue which enqueued asynchronous malloc commands is synchronized
+with, i.e. when `queue::wait` or `queue::wait_and_throw` is called, the SYCL RT
+will check the current total memory allocated to the memory pool against the
+memory pool's deallocation threshold. If the total memory allocated is larger
+than than the threshold then the memory will be freed back to the source,
+otherwise it will not.
+
+[_Note:_ An implementation should maintain memory allocated to a memory pool to
+the size specified by the deallocation threshold when possible, however, an
+implementation is permitted to release this memory back to the source if it is
+required by another memory pool or an application in another process.
+_{endnote}_]
+
+
+== Implementation notes
+
+It is expected that for L0 this extension will be implemented within the L0
+adapter, by reserving allocations for the memory pool and opportunistically
+re-using the memory allocated based on the command lists being enqueued to the
+L0 driver.
+
+It is expected that for CUDA this extension will be implemented by mapping onto
+the CUDA stream-ordered allocator feature.
+
+
+== Issues
+
+. Should we allow mixing asynchronous and synchornous memory commands?
++
+--
+*UNRESOLVED*: CUDA allows memory allocated with the asynchronous malloc command
+to be freed with the regular synchronous free command, should we extend this
+capability to SYCL?
+--
+
+. Should we allow freeing memory with a different queue?
++
+--
+*UNRESOLVED*: Should we allow a memory allocation allocated with an asynchronous
+malloc command from one queue to be freed by an asynchronous free command from
+another queue?
+--
+
+. How should we treat an asynchronous malloc command with a memory pool that is
+not associated with that queue?
```

Comment:
I don't understand this issue.  A memory pool is associated with a context, not with a queue.  Did you mean to ask about the case when the memory pool's context does not match the context of the queue?

Reply:
I've added clarification that `async_malloc` and `async_malloc_from_pool` require the queue to be associated with the same context and devices as the memory pool in https://github.com/intel/llvm/commit/48493ace0babceba6b65fe7399b43284c10681db, so I've removed this issue.